expression/builtin_other.go

```diff
@@ -145,6 +145,17 @@ func (c *inFunctionClass) getFunction(ctx sessionctx.Context, args []Expression)
 		sig.setPbCode(tipb.ScalarFuncSig_InDuration)
 	case types.ETJson:
 		sig = &builtinInJSONSig{baseBuiltinFunc: bf}
+		for i := range bf.args {
+			arg := bf.args[i]
+			// if we need to implicitly cast string to JSON, we only make it a string literal in JSON, instead of parsing it.
+			switch arg.(type) {
+			case *ScalarFunction:
+				f := arg.(*ScalarFunction).Function
```

Comment:
Can we do this in `WrapWithCastAsJSON`?

Reply:
Seems not, it might break things like insert string as json, which needs to parse the string.

Reply:
It seems that `insert` will not invoke the `WrapCastAsxx` funcs.

Reply:
Actually in https://github.com/pingcap/tidb/pull/21903/commits/aafc91327bc7720e13a5d4cbb1b40ae5e709f3e7 the first version, I did this at WrapCastAsJSON, but it breaks the test.

Reply:
Okay, after a more careful research, the failed test is not because of `insert`, it is because the function `json_type` will parse the string, which used `WrapCastAsJSON` to cast string to JSON.

Reply:
I think "not parse json string" is a special case for comparing json with other data types, therefore the current change is appropriate.

Reply:
other json functions require parsing:

* json_valid
* json_type
* JSON_LENGTH
* JSON_DEPTH
* JSON_KEYS
so on so forth, most of json functions require parsing json.